packages/muelu/src/Transfers/GeneralGeometric/MueLu_GeneralGeometricPFactory_def.hpp

```diff
@@ -0,0 +1,843 @@
+	  end_rate[i] = coarse_rate[i];
+	} else {
+	  neH[i] = (neh[i]/coarse_rate[i])+1;
+	}
+      } else {
+	//coarse_rate[i]=1;
+	neh[i]=0;
+	end_rate[i]=1;
+	neH[i]=0;
+      }
+	
+	
+      if(i<ndm) {
+	// From all this data we can compute the number of coarse points NCpts and the number of coefficients in the prolongation matrix.
+	// 1D global  o---x---x ... x---x---+--o
+	// left "o" has coarse_rate term
+	// each "x" has 2*coarse_rate-1 terms and there are (neH+1)-3 "x"
+	// "+" has coarse_rate+end_rate-1 terms
+	// right "o" has end_rate terms
+	//
+	// So the grand total is Nterms = coarse_rate + ((neH+1)-3)*(2*coarse_rate-1) + coarse_rate+end_rate-1 + end_rate
+	//                              = (neH-1)*(2*coarse_rate-1) + 2*end_rate
+	nCpts=nCpts*(neH[i]+1);
+	nTerms=nTerms*((neH[i]-1)*(2*coarse_rate[i]-1)+2*end_rate[i]);
+      }
+	
+      // if(i<ndm) {
+      // 	out << "Direction: " << i << ", coarsening rate: " << coarse_rate[i] << ", adjustement rate: " << end_rate[i] << ", fine elements: " << neh[i] << ", coarse elements: " << neH[i] << ", fine nodes: " << neh[i]+1 << ", coarse nodes: " << neH[i]+1 << std::endl;
+      // } else {
+      // 	out << "Direction: " << i << ", coarsening rate: " << coarse_rate[i] << ", adjustement rate: " << end_rate[i] << ", fine elements: " << neh[i] << ", coarse elements: " << neH[i] << ", fine nodes: " << neh[i] << ", coarse nodes: " << neH[i] << std::endl;
+      // }
+    }
+    // For systems of PDEs we assume that all dofs have the same P operator.
+    nTerms=nTerms*blkSize;
+    // out << "Block size: " << blkSize << std::endl;
+    // out << "Total number of Coarse points: " << nCpts << std::endl;
+    // out << "Number of terms in the prolongation matrix: " << nTerms << std::endl;
+
+    // All that is left to do is loop over NCpts and:
+    //   - extract coarse points coordiate for coarseCoords
+    //   - get coordinates for current stencil computation
+    //   - compute current stencil
+    //   - compute row and column indices for stencil entries
+    LO cNodePerDir[3];
+    LO fNodePerDir[3];
+    for(LO i=0; i<3; ++i) {
+      cNodePerDir[i] = neH[i]+1;
+      fNodePerDir[i] = neh[i]+1;
+    }
+    RCP<const Map> theCoarseMap;
+    RCP<Matrix>    P;
+    LO Ncoarse = MakeGeneralGeometricP(ndm,fNodePerDir,cNodePerDir,coarse_rate,end_rate,
+				       fineCoords, nTerms, blkSize, theCoarseMap, A, P,
+				       coarseCoords);
+
+    // set StridingInformation of P   <---- what is happening here?
+    if (A->IsView("stridedMaps") == true)
+      P->CreateView("stridedMaps", A->getRowMap("stridedMaps"), theCoarseMap);
+    else
+    P->CreateView("stridedMaps", P->getRangeMap(), theCoarseMap);
+    
+    //Xpetra::IO<SC,LO,GO,NO>::Write("Poperator.m",*P);
+
+    // store the transfer operator and node coordinates on coarse level
+    Set(coarseLevel, "P", P);
+    // Set(coarseLevel, "Coordinates", coarseCoords); // <-- for some reason this does not work yet...
```

Comment:
We have to break through a circular dependency: The PFactory cannot request data that it produces.